improver_tests/calibration/dataframe_utilities/test_dataframe_utilities.py

```diff
@@ -397,6 +438,86 @@ def test_basic(self):
+        self.assertCubeEqual(result[1], self.expected_period_truth_threshold)
+
+    def test_realization(self):
+        """Test the expected realization cubes are generated from the input dataframes."""
+        result = forecast_and_truth_dataframes_to_cubes(
+            self.forecast_df_realization,
+            self.truth_subset_df,
+            self.cycletime,
+            self.forecast_period,
+            self.training_length,
+        )
+        self.assertEqual(len(result), 2)
+        self.assertCubeEqual(result[0], self.expected_period_forecast)
+        self.assertCubeEqual(result[1], self.expected_period_truth)
+
+    def test_error_column_missing(self):
+        """Test that an error is raised if dataframe does not contain one of
+        REPRESENTATION_COLUMNS."""
+        msg = "None of the columns"
+        with self.assertRaisesRegex(ValueError, msg):
+            forecast_and_truth_dataframes_to_cubes(
+                self.forecast_df.drop(columns=["percentile"]),
+                self.truth_subset_df,
+                self.cycletime,
+                self.forecast_period,
+                self.training_length,
+            )
+
+    def test_error_multiple_columns(self):
+        """Test that an error is raised if dataframe contains more than one of
+        REPRESENTATION_COLUMNS."""
+        msg = "More than one column"
+        df = self.forecast_df.copy()
+        df["realization"] = 0
+        with self.assertRaisesRegex(ValueError, msg):
+            forecast_and_truth_dataframes_to_cubes(
+                df,
+                self.truth_subset_df,
+                self.cycletime,
+                self.forecast_period,
+                self.training_length,
+            )
+
+    def test_station_id(self):
+        """Test the expected cubes are generated from the input dataframes
+        when inputs contain station_id."""
+        forecast_df = self.forecast_df.copy()
+        forecast_df = forecast_df.loc[forecast_df["wmo_id"].isin(self.wmo_ids[:-1])]
+        forecast_df["station_id"] = forecast_df["wmo_id"].copy()
+        truth_df = self.truth_subset_df.copy()
+        truth_df = truth_df.loc[truth_df["wmo_id"].isin(self.wmo_ids[1:])]
+        truth_df["station_id"] = truth_df["wmo_id"].copy()
+        site_id_values = np.array([0], dtype=np.int32)
+        expected_forecast = self.expected_period_forecast[:, :, [1]].copy()
+        expected_forecast.coord("spot_index").points = site_id_values
+        expected_truth = self.expected_period_truth[:, [1]].copy()
+        expected_truth.coord("spot_index").points = site_id_values
+        result = forecast_and_truth_dataframes_to_cubes(
+            forecast_df,
+            truth_df,
+            self.cycletime,
+            self.forecast_period,
+            self.training_length,
+        )
+        self.assertEqual(len(result), 2)
+        self.assertCubeEqual(result[0], expected_forecast)
```

Comment:
I'm perhaps lacking context for the PR, but I'm slightly confused that although this PR provides support for a `station_id` column in the input dataframes, the resulting cubes only contain a WMO ID coordinate (not a `station_id` coordinate). How would sites without WMO IDs be supported, once converted into cubes, in this situation (assuming that the purpose of the `station_id` column is to support sites withing a WMO ID)?

e.g. if this test was as shown below, the WMO IDs for all the sites would be 0 and there isn't a `station_id` coordinate present to distinguish the sites.

```
        forecast_df = self.forecast_df.copy()
        forecast_df["station_id"] = forecast_df["wmo_id"].copy()
        forecast_df["wmo_id"] = 0
        truth_df = self.truth_subset_df.copy()
        truth_df["station_id"] = truth_df["wmo_id"].copy()
        truth_df["wmo_id"] = 0
        expected_forecast = self.expected_period_forecast.copy()
        expected_truth = self.expected_period_truth.copy()
        result = forecast_and_truth_dataframes_to_cubes(
            forecast_df,
            truth_df,
            self.cycletime,
            self.forecast_period,
            self.training_length,
        )
        self.assertEqual(len(result), 2)
        self.assertCubeEqual(result[0], expected_forecast)
        self.assertCubeEqual(result[1], expected_truth)
```

Reply:
This works ok for our current purposes, since `build_spotdata_cube` will still add a spot_index dimension (with integer labels from 0 to number of rows in the dataframe), so the output still has the right shape. It's true that we lose the station_id information, but later we are using `AggregateReliabilityTables` to combine all the sites, so this doesn't matter. However for clarity and future extensibility, I've now changed this so the station_id data is passed into `build_spotdata_cube` using the `unique_site_id` attribute, so station id will be added as an aux coord in the output.

Reply:
> I've now changed this so the station_id data is passed into build_spotdata_cube using the unique_site_id attribute, so station id will be added as an aux coord in the output.

This behaviour sounds good, such as if used with a calibration method that doesn't aggregate/combine all the sites in the future. It looks like the test code above/below checks that the `unique_site_identifier` attribute is being created.